src/lib/profiles/data-management/Current/GenericTraitCatalogImpl.h

```diff
@@ -0,0 +1,113 @@
+
+namespace nl {
+namespace Weave {
+namespace Profiles {
+namespace WeaveMakeManagedNamespaceIdentifier(DataManagement, kWeaveManagedNamespaceDesignation_Current) {
+/**
+ *  @class GenericTraitCatalogImpl
+ *
+ *  @brief A Weave provided implementation of the TraitCatalogBase interface for a collection of trait data instances
+ *         that all refer to the same resource. It provides a c++ map-backed storage for these instances.
+ */
+template <typename T>
+class GenericTraitCatalogImpl : public TraitCatalogBase<T>
+{
+public:
+    GenericTraitCatalogImpl(void);
+    virtual ~GenericTraitCatalogImpl(void);
+
+    void SetNodeId(uint64_t aNodeId);
+
+    WEAVE_ERROR Add(const ResourceIdentifier & aResourceId, const uint64_t & aInstanceId, PropertyPathHandle basePathHandle,
+                    T * traitInstance, TraitDataHandle & aHandle);
+    WEAVE_ERROR Remove(T * traitInstance);
+    WEAVE_ERROR Remove(TraitDataHandle aHandle);
+    WEAVE_ERROR PrepareSubscriptionSpecificPathList(TraitPath * pathList, uint16_t pathListSize, uint16_t & pathListLen,
+                                                    TraitDataHandle aHandle);
+    WEAVE_ERROR PrepareSubscriptionPathList(TraitPath * pathList, uint16_t pathListSize, uint16_t & pathListLen);
+    WEAVE_ERROR Clear(void);
+
+    virtual WEAVE_ERROR AddressToHandle(TLV::TLVReader & aReader, TraitDataHandle & aHandle,
+                                        SchemaVersionRange & aSchemaVersionRange) const;
+    virtual WEAVE_ERROR HandleToAddress(TraitDataHandle aHandle, TLV::TLVWriter & aWriter,
+                                        SchemaVersionRange & aSchemaVersionRange) const;
+    virtual WEAVE_ERROR Locate(TraitDataHandle aHandle, T ** aTraitInstance) const;
+    virtual WEAVE_ERROR Locate(T * aTraitInstance, TraitDataHandle & aHandle) const;
+
+    WEAVE_ERROR Locate(uint32_t aProfileId, uint64_t aInstanceId, ResourceIdentifier aResourceId, TraitDataHandle & aHandle) const;
+
+    WEAVE_ERROR Locate(uint32_t aProfileId, uint64_t aInstanceId, ResourceIdentifier aResourceId, T ** aTraitInstance) const;
+
+    virtual WEAVE_ERROR DispatchEvent(uint16_t aEvent, void * aContext) const;
+    virtual void Iterate(IteratorCallback aCallback, void * aContext);
+#if WEAVE_CONFIG_ENABLE_WDM_UPDATE
+    virtual WEAVE_ERROR GetInstanceId(TraitDataHandle aHandle, uint64_t & aInstanceId) const;
+    virtual WEAVE_ERROR GetResourceId(TraitDataHandle aHandle, ResourceIdentifier & aResourceId) const;
+#endif // WEAVE_CONFIG_ENABLE_WDM_UPDATE
+
+    /**
+     * Return the number of trait instances in the catalog.
+     */
+    uint32_t Size(void) const;
+
+private:
+    struct CatalogItem
+    {
+        uint32_t mProfileId;
+        uint64_t mInstanceId;
+        ResourceIdentifier mResourceId;
+        T * mItem;
+        PropertyPathHandle mBasePathHandle;
+    };
+
+    uint64_t mNodeId;
+    TraitDataHandle mNextHandle;
+
+    typedef std::map<TraitDataHandle, CatalogItem *> Handle2ItemMap_t;
```

Comment:
Seems superfluous, suggest removing.

Reply:
after removing typedef, ‘Handle2ItemMap_t’ is not a class, namespace, or enumeration

Reply:
Done, thanks